src/main/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProvider.java

```diff
@@ -348,6 +352,26 @@ private AwsKmsMrkAwareMasterKeyProvider(
     }
 
     this.regionalClientSupplier_ = supplier;
+    // Null Pointer Exception (NPE) would be thrown
+    // via result of extractRegion passed to KMS Client request around line 550.
+    // We could refactor extractRegion to handle Strict and Discovery separately,
+    // which would allow us to avoid this potentially un-needed new
+    // DefaultAwsRegionProviderChain().getRegion()
+    // call.
+    // Otherwise, we have to have the call.
+    if (defaultRegion == null) {
+      try {
+        defaultRegion = new DefaultAwsRegionProviderChain().getRegion();
+        if (defaultRegion == null) {
+          throw new AwsCryptoException(
+              "Default Region was not set and Default AWS Region Provider Chain could not load one.");
+        }
+      } catch (SdkClientException ex) {
+        throw new AwsCryptoException(
+            "Default Region was not set and Default AWS Region Provider Chain could not load one.",
+            ex);
+      }
+    }
```

Comment:
I think we should move this behaviour before we check for non ARN's KeyIds either in Builder or before the check. Otherwise, we would still expect exception being thrown.

Reply:
I do not follow. 
Doesn't this check have to happen AFTER the non ARN Key IDs?
Or the illegal combination of "no region with non ARN" cannot be detected.

I believe the intention is to prevent users from providing non ARN Key IDs if the region is not set.
That is what the Unit tests assert.

Reply:
I guess you're correct. 

However, there are [other unit tests](https://github.com/aws/aws-encryption-sdk-java/blob/e94ee8538501f270138feab0a6138bed7ce640ea/src/test/java/com/amazonaws/encryptionsdk/kmssdkv2/AwsKmsMrkAwareMasterKeyProviderTest.java#L489-L507) that check if the default Region is not configured and are passing non-ARN keys. It should resolve with the AWS SDK region. I believe the original changes were asserted such that [defaultRegion == AWS SDK region](https://github.com/aws/aws-encryption-sdk-java/blob/e94ee8538501f270138feab0a6138bed7ce640ea/src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsMrkAwareMasterKeyProvider.java#L83). As a result, if defaultMrkRegion is not configured, it will resolve to defaultRegion, and if the AWS SDK Region cannot be fetched, we expect a failure.

Reply:
I think this code has evolved to be convoluted.
I do not know what the purpose of these conflicting Unit Tests are trying to yield...

I agree, that the current deployed code behaves as you described it:
"An inferred Region is sufficient to allow for non ARNs", 
at least for STRICT, which is what `other_basic_uses` tests.

I am not sure what the behavior is supposed to be for Discovery,
but I... I guess it is the same: "inferred is sufficient if not provided".

So I was entirely wrong about "no region with non ARN" being an illegal combination.

If we set the region in the Builder, we will need to refactor the unit tests to mock the EC2 data loader to confirm that a null returned there fails the build.